x-pack/filebeat/input/httpjson/config_auth.go

```diff
@@ -211,7 +211,7 @@ func (o *oAuth2Config) Validate() error {
 	case oAuth2ProviderGoogle:
 		return o.validateGoogleProvider()
 	case oAuth2ProviderDefault:
-		if o.TokenURL == "" || o.ClientID == "" || o.ClientSecret == "" {
+		if o.TokenURL == "" || o.ClientID == "" {
```

Comment:
This change troubles me. The RFC says that the secret is REQUIRED but that it may be omitted if it's the empty string. This does that, but it allows a user to omit the var completely without error. I would like to have a check that the var is set, even if it is set to `""`.

Reply:
The type of ClientSecret is string therefore the inital value is always "".  
Do you have any idea to check that this var was set?

Reply:
The normal approach would be to make the field be a pointer to the value, so a `*string`, but that will complicate some of the logic.

Reply:
Okay, thx for this hint but I think that going to make it more complicated and needs deeper changes within the logic.
It is an option to set the field as defaulted with `""` in the context of beat?

Reply:
I'll have a look into how complicated it will be to make it aware of set state without using `""` as the zero value.

> It is an option to set the field as defaulted with `""` in the context of beat?

I'm sorry, I don't understand this question.

Reply:
> I'll have a look into how complicated it will be to make it aware of set state without using "" as the zero value.

Thx in advance!
> It is an option to set the field as defaulted with "" in the context of beat?

Sry, my bad. I meant it this way, to declare that the client secret is still required but has the default value `""` if not set.

Reply:
> to declare that the client secret is still required but has the default value `""` if not set.

That is not possible. An unset string is `""`. This is why the pointer is needed. The zero value of a `*string` is `nil`, but the zero value of a `string` is `""`.

Reply:
The following change works (includes tests).
```diff
diff --git a/x-pack/filebeat/input/httpjson/config_auth.go b/x-pack/filebeat/input/httpjson/config_auth.go
index 71911cd6f4..d4e7c4ab5c 100644
--- a/x-pack/filebeat/input/httpjson/config_auth.go
+++ b/x-pack/filebeat/input/httpjson/config_auth.go
@@ -84,7 +84,7 @@ type oAuth2Config struct {
 
        // common oauth fields
        ClientID       string              `config:"client.id"`
-       ClientSecret   string              `config:"client.secret"`
+       ClientSecret   *string             `config:"client.secret"`
        EndpointParams map[string][]string `config:"endpoint_params"`
        Password       string              `config:"password"`
        Provider       oAuth2Provider      `config:"provider"`
@@ -114,7 +114,7 @@ func (o *oAuth2Config) isEnabled() bool {
 func (o *oAuth2Config) clientCredentialsGrant(ctx context.Context, _ *http.Client) *http.Client {
        creds := clientcredentials.Config{
                ClientID:       o.ClientID,
-               ClientSecret:   o.ClientSecret,
+               ClientSecret:   maybeString(o.ClientSecret),
                TokenURL:       o.getTokenURL(),
                Scopes:         o.Scopes,
                EndpointParams: o.getEndpointParams(),
@@ -131,10 +131,10 @@ func (o *oAuth2Config) client(ctx context.Context, client *http.Client) (*http.C
                if o.User != "" || o.Password != "" {
                        conf := &oauth2.Config{
                                ClientID:     o.ClientID,
-                               ClientSecret: o.ClientSecret,
+                               ClientSecret: maybeString(o.ClientSecret),
                                Endpoint: oauth2.Endpoint{
                                        TokenURL:  o.TokenURL,
-                                       AuthStyle: authStyleInParams,
+                                       AuthStyle: oauth2.AuthStyleAutoDetect,
                                },
                        }
                        token, err := conf.PasswordCredentialsToken(ctx, o.User, o.Password)
@@ -167,6 +167,14 @@ func (o *oAuth2Config) client(ctx context.Context, client *http.Client) (*http.C
        }
 }
 
+// maybeString returns the string pointed to by p or "" if p in nil.
+func maybeString(p *string) string {
+       if p == nil {
+               return ""
+       }
+       return *p
+}
+
 // getTokenURL returns the TokenURL.
 func (o *oAuth2Config) getTokenURL() string {
        switch o.getProvider() {
@@ -211,7 +219,7 @@ func (o *oAuth2Config) Validate() error {
        case oAuth2ProviderGoogle:
                return o.validateGoogleProvider()
        case oAuth2ProviderDefault:
-               if o.TokenURL == "" || o.ClientID == "" || o.ClientSecret == "" {
+               if o.TokenURL == "" || o.ClientID == "" || o.ClientSecret == nil {
                        return errors.New("both token_url and client credentials must be provided")
                }
                if (o.User != "" && o.Password == "") || (o.User == "" && o.Password != "") {
@@ -228,7 +236,7 @@ func (o *oAuth2Config) Validate() error {
 var findDefaultGoogleCredentials = google.FindDefaultCredentials
 
 func (o *oAuth2Config) validateGoogleProvider() error {
-       if o.TokenURL != "" || o.ClientID != "" || o.ClientSecret != "" ||
+       if o.TokenURL != "" || o.ClientID != "" || o.ClientSecret != nil ||
                o.AzureTenantID != "" || o.AzureResource != "" || len(o.EndpointParams) != 0 {
                return errors.New("none of token_url and client credentials can be used, use google.credentials_file, google.jwt_file, google.credentials_json or ADC instead")
        }
@@ -295,7 +303,7 @@ func (o *oAuth2Config) validateAzureProvider() error {
        if o.TokenURL != "" && o.AzureTenantID != "" {
                return errors.New("only one of token_url and tenant_id can be used")
        }
-       if o.ClientID == "" || o.ClientSecret == "" {
+       if o.ClientID == "" || o.ClientSecret == nil {
                return errors.New("client credentials must be provided")
        }
 
diff --git a/x-pack/filebeat/input/httpjson/config_test.go b/x-pack/filebeat/input/httpjson/config_test.go
index 2ee8c4276c..116ea617e3 100644
--- a/x-pack/filebeat/input/httpjson/config_test.go
+++ b/x-pack/filebeat/input/httpjson/config_test.go
@@ -152,6 +152,32 @@ func TestConfigOauth2Validation(t *testing.T) {
                                "auth.oauth2": map[string]interface{}{},
                        },
                },
+               {
+                       name: "client credential secret may be empty",
+                       input: map[string]interface{}{
+                               "auth.oauth2": map[string]interface{}{
+                                       "enabled":   true,
+                                       "token_url": "localhost",
+                                       "client": map[string]interface{}{
+                                               "id":     "a_client_id",
+                                               "secret": "",
+                                       },
+                               },
+                       },
+               },
+               {
+                       name:        "client credential secret may not be missing",
+                       expectedErr: "both token_url and client credentials must be provided accessing 'auth.oauth2'",
+                       input: map[string]interface{}{
+                               "auth.oauth2": map[string]interface{}{
+                                       "enabled":   true,
+                                       "token_url": "localhost",
+                                       "client": map[string]interface{}{
+                                               "id": "a_client_id",
+                                       },
+                               },
+                       },
+               },
                {
                        name: "if user and password is set oauth2 must use user-password authentication",
                        input: map[string]interface{}{
```

Reply:
Thank you very much! I've made the changes above.